File: src/algo/refill.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   refill.c                                           :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: vpolard <vpolard@student.42.fr>            +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2026/03/02 15:48:10 by vpolard           #+#    #+#             */
/*   Updated: 2026/03/09 12:13:36 by vpolard          ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "algo.h"

static int	find_target_value(t_stack *stack_a, int value_to_push)
{
	int	index;
	int	next_value;

	index = 0;
	while (index <= stack_a->btm)
	{
		if (index == stack_a->btm)
			next_value = stack_a->content[0];
		else
			next_value = stack_a->content[index + 1];
		if ((stack_a->content[index] < value_to_push
				&& value_to_push < next_value) || (stack_a->content[index]
				> next_value && (value_to_push > stack_a->content[index]
					|| value_to_push < next_value)))
			return (next_value);
		index++;
	}
	return (stack_a->content[0]);
}

static int	find_target_index(t_stack *stack_a, int target_value)
{
	int	index;

	index = 0;
	while (index <= stack_a->btm)
	{
		if (stack_a->content[index] == target_value)
			return (index);
		index++;
	}
	return (0);
}

void	refill(t_package *package)
{
	int	value;
	int	target_index;
	int	target_value;

	while (package->b->btm >= 0)
	{
		value = package->b->content[0];
		target_value = find_target_value(package->a, val);
		target_index = find_target_index(package->a, t_val);
		while (package->a->content[0] != target_value)
		{
			if (target_index <= package->a->btm / 2)
				ra(package);
			else
				rra(package);
		}
		pa(package);
	}
}

File: rust/src/algo/refill.rs
use super::Package;

impl Package {
    pub fn refill(&mut self) -> &mut Self {
        let mut target_value: i32;
        let mut target_index: usize;
        while self.b.content.len() > 0 {
            target_value = self.find_value(self.b.content[0]);
            target_index = self.find_index(target_value);
            while self.a.content[0] != target_value {
                if target_index <= (self.a.content.len() - 1) / 2 {
                    self.ra();
                } else {
                    self.rra();
                }
            }
            self.pa();
        }
        self
    }
    fn find_value(&self, value: i32) -> i32 {
        let mut next_value: i32;
        for (index, element) in self.a.content.iter().enumerate() {
            if index == self.a.content.len() - 1 {
                next_value = self.a.content[0];
            } else {
                next_value = self.a.content[index + 1];
            }
            if (*element < value && value < next_value)
                || *element > next_value && (value > *element || value < next_value)
            {
                return next_value;
            }
        }
        return self.a.content[0];
    }
    fn find_index(&self, value: i32) -> usize {
        for (index, element) in self.a.content.iter().enumerate() {
            if *element == value {
                return index;
            }
        }
        0
    }
}
